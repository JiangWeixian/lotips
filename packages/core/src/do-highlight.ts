/**
 * @description replace `highlight` -> `<mark>highlight</mark>` with with vanilla js.
 * @see: {@link https://www.eriwen.com/javascript/highlight-search-results-with-js/}
 * @category string
 * @example doHighlight('hello world', 'world')
 */
export const doHighlight = (
  original: string,
  highlight: string,
  callback?: (substring: string, ...args: any[]) => string,
) => {
  if (!highlight) return original

  if (callback)
    return original.replace(new RegExp(`>([^<]*)?(${highlight})([^>]*)?<`, 'ig'), callback)

  return original.replace(
    new RegExp(`>([^<]*)?(${highlight})([^>]*)?<`, 'ig'),
    '>$1<mark>$2</mark>$3<',
  )
}
