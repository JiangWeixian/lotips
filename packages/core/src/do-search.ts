/**
 * @fileoverview a simple way to highlight content on website
 * @waring now it's not working on textarea
 */

// tslint:disable:no-any
export function doSearch(
  text: string,
  { aCaseSensitive = false }: { aCaseSensitive: boolean } = { aCaseSensitive: false },
) {
  const W = window as any
  const B = document.body as any
  if (W && window.getSelection) {
    document.designMode = 'on'
    const sel = window.getSelection()
    sel?.collapse(document.body, 0)

    // tslint:disable-next-line:no-any
    while (W.find(text, aCaseSensitive)) {
      document.execCommand('HiliteColor', false, 'yellow')
      sel?.collapseToEnd()
    }
    document.designMode = 'off'
  } else if (B.createTextRange) {
    const textRange = B.createTextRange()
    while (textRange.findText(text)) {
      textRange.execCommand('BackColor', false, 'yellow')
      textRange.collapse(false)
    }
  }
}
