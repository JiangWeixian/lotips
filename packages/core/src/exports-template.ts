import uniq from 'lodash-es/uniq'
import isEmpty from 'lodash-es/isEmpty'

export type ExportsTemplateParams = {
  /**
   * Exported module names
   */
  names: string[]
  /**
   * Exported formats, impact `exports.require` or `exports.import`
   * @default ['cjs', 'esm']
   */
  formats?: Array<'cjs' | 'esm' | 'es'>
  /**
   * Bundled files dir
   * @default dist
   */
  dirs?:
    | string
    | {
        cjs: string
        esm: string
      }
  /**
   * Config cjs format and esm format ext
   * @default { cjs: 'cjs', esm: 'mjs' }
   */
  exts?:
    | string
    | {
        cjs: string
        esm: string
      }
  /**
   * Enable types define, will setup `typesVersions` and `exports.types`
   */
  types?:
    | boolean
    | {
        dir: string
      }
}

type PartPkg = {
  exports: Record<string, Record<string, string> | string>
  main?: string
  module?: string
  types?: string
  typesVersions?: Record<string, Record<string, string[]>>
}

/**
 * @description generate package.json exports/typesVersions/main/module/types field
 * @category template
 * @example exportsTemplate({
 *  names: ['index', 'ua', 'do-something/index']
 * })
 */
export const exportsTemplate = ({
  names,
  formats = ['cjs', 'esm'],
  dirs = 'dist',
  types = true,
  exts = { cjs: 'cjs', esm: 'mjs' },
}: ExportsTemplateParams) => {
  const pkg: PartPkg = {
    exports: {},
  }
  // resolve options
  const _formats = uniq(
    formats.map(format => {
      if (format === 'es') {
        return 'esm'
      }
      return format
    }),
  )
  const _dirs =
    typeof dirs === 'string'
      ? {
          cjs: dirs,
          esm: dirs,
        }
      : dirs
  const _exts =
    typeof exts === 'string'
      ? {
          cjs: exts,
          esm: exts,
        }
      : exts
  const _types =
    typeof types === 'boolean'
      ? {
          dir: _dirs.cjs,
        }
      : undefined
  pkg.exports['./package.json'] = './package.json'
  if (types) {
    pkg.typesVersions = {
      '*': {},
    }
  }
  const subExport = (name: string) => {
    const results: Record<string, string> = {}
    _formats.forEach(format => {
      if (format === 'cjs') {
        results.require = `./${_dirs.cjs}/${name}.${_exts.cjs}`
      }
      if (format === 'cjs') {
        results.import = `./${_dirs.esm}/${name}.${_exts.esm}`
      }
      if (_types) {
        results.types = `./${_types.dir}/${name}.d.ts`
      }
    })
    return results
  }
  const prettyName = (name: string) => {
    if (name.endsWith('/index')) {
      return name.slice(0, -6)
    }
    return name
  }
  names.forEach(name => {
    if (name === 'index') {
      pkg.exports['.'] = subExport('index')
      pkg.main = pkg.exports['.'].require.slice(2)
      pkg.module = pkg.exports['.'].import.slice(2)
      if (types) {
        pkg.types = pkg.exports['.'].types.slice(2)
      }
    } else {
      const _name = prettyName(name)
      pkg.exports[`./${_name}`] = subExport(name)
      if (_types) {
        pkg.typesVersions!['*'][_name] = [
          (pkg.exports[`./${_name}`] as Record<string, string>).types.slice(2),
        ]
      }
    }
  })
  if (isEmpty(pkg.typesVersions!['*'])) {
    delete pkg.typesVersions
  }
  return pkg
}
