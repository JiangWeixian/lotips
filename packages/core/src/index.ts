export { isMobile, isiOS, isAndroid } from './ua'

export { doHighlight } from './do-highlight'
export { doSearch } from './do-search'

export { exportsTemplate } from './exports-template'
