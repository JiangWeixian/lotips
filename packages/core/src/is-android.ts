// refs: https://github.com/smali-kazmi/detect-mobile-browser/blob/master/detect-browser.js
export const isAndroid = (ua: string = navigator.userAgent) => {
  return ua.match(/Android/i)
}
