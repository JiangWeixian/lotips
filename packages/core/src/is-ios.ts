// refs: https://github.com/smali-kazmi/detect-mobile-browser/blob/master/detect-browser.js
export const isiOS = (ua: string = navigator.userAgent) => {
  return ua.match(/iPhone|iPad|iPod/i)
}
