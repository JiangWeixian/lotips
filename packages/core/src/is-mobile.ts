// refs: https://github.com/juliangruber/is-mobile/blob/master/index.js
import IsMobile, { IsMobileOptions } from 'is-mobile'

export const isMobile = (props: IsMobileOptions = {}) => {
  return IsMobile(props)
}
