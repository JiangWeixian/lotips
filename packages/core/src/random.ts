/**
 * @description return -1 or 1
 */
export const direction = () => {
  return Math.random() < 0.5 ? -1 : 1
}
