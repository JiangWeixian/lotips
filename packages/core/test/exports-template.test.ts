import { describe, it, expect } from 'vitest'

import { exportsTemplate } from '../src/exports-template'

describe('exports template', () => {
  it('demo part', () => {
    const exports = exportsTemplate({
      names: ['index', 'ua', 'do-something/index']
    })
    expect(exports).toMatchSnapshot('basic')
  })
})
