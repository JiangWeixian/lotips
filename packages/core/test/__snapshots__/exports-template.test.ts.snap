// Vitest Snapshot v1

exports[`exports template > basic > basic 1`] = `
{
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs",
      "types": "./dist/index.d.ts",
    },
    "./do-something": {
      "import": "./dist/do-something/index.mjs",
      "require": "./dist/do-something/index.cjs",
      "types": "./dist/do-something/index.d.ts",
    },
    "./package.json": "./package.json",
    "./ua": {
      "import": "./dist/ua.mjs",
      "require": "./dist/ua.cjs",
      "types": "./dist/ua.d.ts",
    },
  },
  "main": "dist/index.cjs",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "typesVersions": {
    "*": {
      "do-something": [
        "dist/do-something/index.d.ts",
      ],
      "ua": [
        "dist/ua.d.ts",
      ],
    },
  },
}
`;

exports[`exports template > non index file, main/module/types filed will be empty > non index 1`] = `
{
  "exports": {
    "./package.json": "./package.json",
    "./ua": {
      "import": "./dist/ua.mjs",
      "require": "./dist/ua.cjs",
      "types": "./dist/ua.d.ts",
    },
  },
  "typesVersions": {
    "*": {
      "ua": [
        "dist/ua.d.ts",
      ],
    },
  },
}
`;

exports[`exports template > only index file > index only 1`] = `
{
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs",
      "types": "./dist/index.d.ts",
    },
    "./package.json": "./package.json",
  },
  "main": "dist/index.cjs",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
}
`;
